[Issue 738] Include OS/Processor Info in UserAgent

dmiluski opened the pull request:
## 🔧 Issue:
https://github.com/mapbox/mobile-telemetry/issues/738

## 🥅 Goal:
Having information about the OS and CPU can help us understand our market segmentation for devices. As well as provide insights into the numbers of users running on simulators or desktop..

## 🗺️  Approach:
Leverage UIDevice / NSProcessInfo / sys info to extract beneficial information

## 👨‍⚕️ Changes:
+ Add NSProcessorInfo category methods to host processor and operating system helpers
+ Include OS/Version in root user agent
+ Include Processor info in user agent

## Before/After
| User Agent (Before) | User Agent (After) |
| --- | --- |
|  MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1) MapboxMobileEvents/0.10.2 (com.mapbox.MapboxMobileEvents; v1) | MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 13.5; ARME) MapboxMobileEvents/0.10.2 (com.mapbox.MapboxMobileEvents; v1) |
| | (Includes extra `iOS 13.5; ARME`) |

dmiluski committed:
[Issue 738] Include OS info in UserAgent

Issue:
https://github.com/mapbox/mobile-telemetry/issues/738

Goal:
User Agent should include OS Version and CPU info

Notes:
Although OS/Version is available through UIDevice, CPU information is not

Approach:
Filing provisionally to see if this is helpful

Changes:

| User Agent (Before) | User Agent (After) |
| --- | --- |
|  MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1) MapboxMobileEvents/0.10.2 (com.mapbox.MapboxMobileEvents; v1) | MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 13.5) MapboxMobileEvents/0.10.2 (com.mapbox.MapboxMobileEvents; v1) |

dmiluski commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -285,12 +285,16 @@ - (NSString *)mme_userAgentString {
     dispatch_once(&onceToken, ^{
 
-        userAgent = [NSString stringWithFormat:@"%@/%@ (%@; v%@)",
+
```
Given this is directly grabbing main bundle and UIDevice, testability here is difficult given variations coming from bundle info.plist which can be changed out from under this code.

Options for boosting testability:
1. Include an additional function which takes Bundle/UIDevice arguments (Potentially another class/interface) to allow for mocking known inputs and potential problematic info for verification.

alfwatt replied:
it uses mme_mainBundle so you can inject that in tests at least

alfwatt replied:
We have `NSProcessInfo+SystemInfo` to use for this

alfwatt commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -304,7 +308,10 @@ - (NSString *)mme_userAgentString {
                 userAgent = [userAgent stringByAppendingString:uaFragment];
             }
         }
+        NSLog(@"Dane - constructed agent: %@",userAgent);
```
Stray log message

dmiluski replied:
Good eye, removed

alfwatt reviewed (changes requested):
We should bring `NSProcessInfo+SystemInfo` over from the beta branch for these, and add a method there for the processor architecture.

misterkerry reviewed (dismissed):
lgtm

dmiluski committed:
[UserAgent] Include CPU Information

Issue:
https://github.com/mapbox/mobile-telemetry/issues/738

Goal:
To gather perspective on what devices are being run, include CPU.

References:
https://opensource.apple.com/source/xnu/xnu-792/libkern/libkern/sysctl.h

Constants
```
/*
 *	ARM subtypes
 */
#define CPU_SUBTYPE_ARM_ALL             ((cpu_subtype_t) 0)
#define CPU_SUBTYPE_ARM_V4T             ((cpu_subtype_t) 5)
#define CPU_SUBTYPE_ARM_V6              ((cpu_subtype_t) 6)
#define CPU_SUBTYPE_ARM_V5TEJ           ((cpu_subtype_t) 7)
#define CPU_SUBTYPE_ARM_XSCALE          ((cpu_subtype_t) 8)
#define CPU_SUBTYPE_ARM_V7              ((cpu_subtype_t) 9)  /* ARMv7-A and ARMv7-R */
#define CPU_SUBTYPE_ARM_V7F             ((cpu_subtype_t) 10) /* Cortex A9 */
#define CPU_SUBTYPE_ARM_V7S             ((cpu_subtype_t) 11) /* Swift */
#define CPU_SUBTYPE_ARM_V7K             ((cpu_subtype_t) 12)
#define CPU_SUBTYPE_ARM_V8              ((cpu_subtype_t) 13)
#define CPU_SUBTYPE_ARM_V6M             ((cpu_subtype_t) 14) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V7M             ((cpu_subtype_t) 15) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V7EM            ((cpu_subtype_t) 16) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V8M             ((cpu_subtype_t) 17) /* Not meant to be run under xnu */

/*
 *  ARM64 subtypes
 */
#define CPU_SUBTYPE_ARM64_ALL           ((cpu_subtype_t) 0)
#define CPU_SUBTYPE_ARM64_V8            ((cpu_subtype_t) 1)
#define CPU_SUBTYPE_ARM64E              ((cpu_subtype_t) 2)
```

Approach:
Inspect CPU via machine C functions which extract constants to map to human readable code.

Expected Output:
eCarra/4.0 (com.ecarra.ios-beta; v1; iOS 13.4.5; ppc64) Mapbox/5.9.0 (com.mapbox.Mapbox; v15256) . . .

Tested Output:
CPU Type (On Device missing processor info)
MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 12.4; x86)

MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 13.4.1; ARME

Changes:
+ Create NSProcessInfo Category to extract CPU info

dmiluski commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -280,16 +283,73 @@ - (NSURL *)mme_configServiceURL {
     return serviceURL;
 }
 
+- (NSString *)cpuType {
```
@alfwatt @rclee 

Initial demo of capturing user readable CPU info.

On initial test I tested on my computer and my phone. Upon testing my phone I received an empty type which upon investigation found that it fell under the x86 64 type.

Although this shows promise, this may now be a liability that needs to be updated for each hardware update yielding an empty string if not falling in these buckets.

Current Output:
```
CPU Type (Tested on corp laptop)
MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 12.4; x86)

CPU Type (Tested on personal iPhone 11 Pro)
MMETestHost/1.0.0 (com.mapbox.MMETestHost; v1; iOS 13.4.1; ARME
```

TODO:
Given an okay on the approach I can migrate this to a more testable solution.

dmiluski commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -285,11 +287,15 @@ - (NSString *)mme_userAgentString {
             NSBundle.mme_mainBundle.bundleIdentifier,
-            NSBundle.mme_mainBundle.infoDictionary[(id)kCFBundleVersionKey]];
+            NSBundle.mme_mainBundle.infoDictionary[(id)kCFBundleVersionKey],
+            UIDevice.currentDevice.systemName,
```
Include newly available system name/version/cpuType

dmiluski commented on MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m:
```diff
@@ -0,0 +1,65 @@
+
+@implementation NSProcessInfo (SystemInfo)
+
++ (NSString*)stringForProcessorType:(NSInteger)type subtype:(NSInteger)subtype {
```
Note:
If type/subtype are not found in this sequence of checks, value will return an empty string ""

alfwatt commented on MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m:
```diff
@@ -0,0 +1,65 @@
+
+@implementation NSProcessInfo (SystemInfo)
+
++ (NSString*)stringForProcessorType:(NSInteger)type subtype:(NSInteger)subtype {
```
We're in a public namespace so `mme_string…`

dmiluski replied:
Agreed. Updated

alfwatt commented on MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m:
```diff
@@ -0,0 +1,65 @@
+        switch(subtype)
+        {
+            case CPU_SUBTYPE_ARM_V6:
+                [cpu appendString:@"V6"];
```
Can we make these `v6` etc so it's `ARMv6` like in their docs: https://static.docs.arm.com/ddi0419/d/DDI0419D_armv6m_arm.pdf

dmiluski replied:
Sure thing.

Did you have any thoughts on the subtype E? I wasn't sure if that meant bionic? Cortext?

dmiluski replied:
I was just comparing with our supported architectures:

`archs:           [ armv7, armv7s, arm64, arm64e ]`

How would you feel about an all lowercased output?

dmiluski replied:
eek, looks like I missed a few more sub type variants:

```
#define CPU_SUBTYPE_ARM_ALL             ((cpu_subtype_t) 0)
#define CPU_SUBTYPE_ARM_V4T             ((cpu_subtype_t) 5)
#define CPU_SUBTYPE_ARM_V6              ((cpu_subtype_t) 6)
#define CPU_SUBTYPE_ARM_V5TEJ           ((cpu_subtype_t) 7)
#define CPU_SUBTYPE_ARM_XSCALE          ((cpu_subtype_t) 8)
#define CPU_SUBTYPE_ARM_V7              ((cpu_subtype_t) 9)  /* ARMv7-A and ARMv7-R */
#define CPU_SUBTYPE_ARM_V7F             ((cpu_subtype_t) 10) /* Cortex A9 */
#define CPU_SUBTYPE_ARM_V7S             ((cpu_subtype_t) 11) /* Swift */
#define CPU_SUBTYPE_ARM_V7K             ((cpu_subtype_t) 12)
#define CPU_SUBTYPE_ARM_V8              ((cpu_subtype_t) 13)
#define CPU_SUBTYPE_ARM_V6M             ((cpu_subtype_t) 14) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V7M             ((cpu_subtype_t) 15) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V7EM            ((cpu_subtype_t) 16) /* Not meant to be run under xnu */
#define CPU_SUBTYPE_ARM_V8M             ((cpu_subtype_t) 17) /* Not meant to be run under xnu */
```

alfwatt commented on MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // Return Copy to Remove Mutability
+    return [cpu copy];
```
Pretty sure that's just another mutable string, you want `[NSString stringWithString:cpu]` for an immutable copy

dmiluski replied:
If called upon NSMutable string it returns an immutable variant. If called with mutableString, it's responded with a mutable variant.

```
    // Return Copy to Remove Mutability
    NSMutableString * scratch = [cpu copy];
    [scratch appendString:@"Foo"]; >>> Results in exception
    
    return [cpu copy];
```
Due to:
`2020-06-17 15:53:31.032713-0700 MMETestHost[61435:5325632] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString appendString:]: unrecognized selector sent to instance 0xefb77832e6209c70'`

Although we could return a mutable string, this was just something I learned as a younger developer to prevent someone mucking with it later.

dmiluski replied:
Re-Reading, the copy isn't really necessary since we aren't retaining a reference. Removing it. Sorry for the confusion.

alfwatt replied:
> If called upon NSMutable string it returns an immutable variant

That's surprising, isn't it `instancetype`?

alfwatt commented on MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m:
```diff
@@ -0,0 +1,65 @@
+    return [cpu copy];
+}
+
++ (NSString *)processorTypeDescription {
```
`mme_`

dmiluski replied:
agreed, updated

alfwatt commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -285,11 +287,15 @@ - (NSString *)mme_userAgentString {
             NSBundle.mme_mainBundle.bundleIdentifier,
-            NSBundle.mme_mainBundle.infoDictionary[(id)kCFBundleVersionKey]];
+            NSBundle.mme_mainBundle.infoDictionary[(id)kCFBundleVersionKey],
+            UIDevice.currentDevice.systemName,
```
[mme_osVersion](https://github.com/mapbox/mapbox-events-ios/blob/7a943206b647b27b736a719f39dacf17f520377e/MapboxMobileEvents/Categories/NSProcessInfo%2BSystemInfo.m#L64) would be nice here to remove the UIDevice dependency

dmiluski replied:
Although I can see the benefit of an abstraction from UIDevice. 

The SDK at present does not support MacOS either as a target. I wonder if it's necessary to get in just before a release? Could this perhaps wait for a follow up release which has the necessary targets to require implementation?

dmiluski replied:
Updated

alfwatt reviewed (changes requested):
Wish we had space for `CPU_SUBTYPE_VAX_ALL`

dmiluski committed:
PR Feedback: Decouple from UIDevice

+ Include operating system descriptor from beta branch

https://github.com/mapbox/mapbox-events-ios/blob/aw-mme-1-0-beta/MapboxMobileEvents/Categories/NSProcessInfo+SystemInfo.m#L64

dmiluski commented on Tests/NSProcessInfoTests+SystemInfo.m:
```diff
@@ -0,0 +1,38 @@
+    XCTAssertEqualObjects([NSProcessInfo mme_stringForProcessorType:CPU_TYPE_X86_64 subtype:CPU_SUBTYPE_X86_ALL], @"x86_64");
+
+    // ARM CPUs
+    XCTAssertEqualObjects([NSProcessInfo mme_stringForProcessorType:CPU_TYPE_ARM subtype:CPU_SUBTYPE_ARM_ALL], @"arm");
```
Opted for lowercase descriptors to match our projects supported architecture types:
` [ armv7, armv7s, arm64, arm64e ]`

This felt a bit easier to read and differentiate values than
`[ ARMv7, ARMv7S, ARM64, ARM64E ]`

alfwatt replied:
looks good!

dmiluski commented:
Hi @alfwatt ,

If you have the time, could you give a followup review given your request for change? I believe I may have satisfied your requests.

alfwatt reviewed (approved):
Thanks!
